File: ui/src/utils/resumeParser.ts
export interface ParsedJob {
  title: string;
  date: string;
  company: string;
  location: string;
  bullets: string[];
  techStack?: string;
}

export interface ParsedEdu {
  institution: string;
  date: string;
  degree: string;
}

export interface ParsedSkill {
  category: string;
  items: string;
}

export interface ParsedCustomSection {
  title: string;
  content: string[];
}

export interface ParsedResume {
  name: string;
  title: string;
  contactItems: string[];
  summary: string;
  skills: ParsedSkill[];
  workExperience: ParsedJob[];
  education: ParsedEdu[];
  customSections: ParsedCustomSection[];
}

export function parseResumeStructure(text: string): ParsedResume {
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
  const result: ParsedResume = {
    name: '',
    title: '',
    contactItems: [],
    summary: '',
    skills: [],
    workExperience: [],
    education: [],
    customSections: []
  };

  if (lines.length === 0) return result;

  // Header extraction
  result.name = lines[0].replace(/\[cite:\s*\d+\]/g, '').trim();
  let idx = 1;

  while (idx < lines.length) {
    const line = lines[idx].replace(/\[cite:\s*\d+\]/g, '').trim();
    const isSectionHeader = /^(PROFESSIONAL SUMMARY|SUMMARY|WORK EXPERIENCES|WORK EXPERIENCE|EXPERIENCE|SKILLS|TECHNICAL SKILLS|EDUCATIONS|EDUCATION|PROJECTS|CERTIFICATIONS)$/i.test(line);
    if (isSectionHeader) break;

    const hasEmail = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/.test(line);
    const hasPhone = /(?:\+?\d{1,3}[\s-]?)?\(?\d{3,5}\)?[\s-]?\d{3,5}[\s-]?\d{3,5}/.test(line);

    if (!result.title && !hasEmail && !hasPhone && !/[|•]/.test(line) && line.length < 80) {
      result.title = line;
    } else {
      let workLine = line;

      // Extract title if glued to contact string
      const titleMatch = workLine.match(/^([A-Za-z\s,\(\)\/-]+?(?:Engineer|Developer|Architect|Manager|Lead|Specialist|Consultant|Scientist|Designer|Analyst|Programmer))\s*(?=\+?\d|[\w.-]+@|[A-Z][a-z]+,|$)/i);
      if (titleMatch && !result.title) {
        result.title = titleMatch[1].trim();
        workLine = workLine.substring(titleMatch[0].length).trim();
      }

      // Extract Email
      const emailMatch = workLine.match(/([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})/);
      if (emailMatch) {
        result.contactItems.push(emailMatch[1]);
        workLine = workLine.replace(emailMatch[1], ' ').trim();
      }

      // Extract Phone
      const phoneMatch = workLine.match(/(\+?\d{1,3}[\s-]?\d{10,12}|\+?\d{1,4}[\s-]?\d{3,5}[\s-]?\d{3,5})/);
      if (phoneMatch) {
        const cleanPhone = phoneMatch[1].replace(/(\+\d{2})(\d{10})/, '$1 $2');
        result.contactItems.push(cleanPhone);
        workLine = workLine.replace(phoneMatch[1], ' ').trim();
      }

      // Remaining contact parts
      const remainingParts = workLine.split(/[|•]/).map(p => p.trim()).filter(Boolean);
      remainingParts.forEach(p => {
        if (p && !result.contactItems.includes(p)) {
          result.contactItems.push(p);
        }
      });
    }
    idx++;
  }

  // Parse Sections
  let currentSection = '';
  let currentJob: ParsedJob | null = null;
  let currentEdu: ParsedEdu | null = null;

  for (; idx < lines.length; idx++) {
    const rawLine = lines[idx];
    const line = rawLine.replace(/\[cite:\s*\d+\]/g, '').trim();
    if (!line) continue;

    const isSectionHeader = /^(PROFESSIONAL SUMMARY|SUMMARY|WORK EXPERIENCES|WORK EXPERIENCE|EXPERIENCE|SKILLS|TECHNICAL SKILLS|EDUCATIONS|EDUCATION|PROJECTS|CERTIFICATIONS)$/i.test(line)
      || (line.length < 40 && line === line.toUpperCase() && !line.includes('|') && !line.includes('@') && !line.includes('+') && line.length > 3);

    if (isSectionHeader) {
      if (currentJob) { result.workExperience.push(currentJob); currentJob = null; }
      if (currentEdu) { result.education.push(currentEdu); currentEdu = null; }

      if (/SUMMARY/i.test(line)) currentSection = 'SUMMARY';
      else if (/SKILL/i.test(line)) currentSection = 'SKILLS';
      else if (/WORK|EXPERIENCE/i.test(line)) currentSection = 'WORK';
      else if (/EDU/i.test(line)) currentSection = 'EDUCATION';
      else {
        currentSection = line;
        result.customSections.push({ title: line, content: [] });
      }
      continue;
    }

    if (currentSection === 'SUMMARY') {
      result.summary = result.summary ? result.summary + ' ' + line : line;
    } else if (currentSection === 'SKILLS') {
      if (line.includes(':')) {
        const [cat, items] = line.split(':', 2);
        const catTrim = cat.trim();
        const itemsTrim = items.trim();
        if (itemsTrim) {
          result.skills.push({ category: catTrim, items: itemsTrim });
        } else {
          result.skills.push({ category: catTrim, items: '' });
        }
      } else if (result.skills.length > 0 && !result.skills[result.skills.length - 1].items) {
        result.skills[result.skills.length - 1].items = line.replace(/^[•\-*▪◦\s]+/, '').trim();
      } else {
        const txt = line.replace(/^[•\-*▪◦\s]+/, '').trim();
        result.skills.push({ category: 'Key Skills', items: txt });
      }
    } else if (currentSection === 'WORK') {
      const fullRangeRx = /((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4})[-–\s\u2013\u2014]+(?:Present|\d{4}|Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec))/i;

      if (line.toLowerCase().startsWith('technologies') || line.toLowerCase().startsWith('technologies / skills used')) {
        if (currentJob) {
          const tech = line.replace(/^technologies\s*\/\s*skills\s*used\s*:\s*/i, '').replace(/^technologies\s*used\s*:\s*/i, '').trim();
          currentJob.techStack = tech;
        }
        continue;
      }

      const fullDateMatch = line.match(fullRangeRx);
      const endsWithDateMatch = line.match(/((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4}))\s*$/i);

      // Handle split date across two lines
      if (currentJob && currentJob.date && !currentJob.date.includes('-') && !currentJob.date.includes('–') && !currentJob.date.includes('Present') && endsWithDateMatch) {
        const endDate = endsWithDateMatch[1];
        currentJob.date = `${currentJob.date} - ${endDate}`;
        const comp = line.replace(/((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4}))\s*$/i, '').trim();
        if (comp) currentJob.company = comp;
        continue;
      }

      // Handle job created with title on prev line, date and company on current line
      if (currentJob && currentJob.title && !currentJob.date && !currentJob.company && fullDateMatch) {
        currentJob.date = fullDateMatch[1];
        currentJob.company = line.replace(fullRangeRx, '').trim().replace(/[\s|–—-]+$/, '').trim();
        continue;
      }

      if (fullDateMatch && line.length < 130 && !/^[•\-*▪◦]/.test(line)) {
        if (currentJob) { result.workExperience.push(currentJob); }
        const date = fullDateMatch[1];
        const title = line.replace(fullRangeRx, '').trim().replace(/[\s|–—-]+$/, '').trim();
        currentJob = { title, date, company: '', location: '', bullets: [] };
      } else if (endsWithDateMatch && line.length < 100 && !currentJob?.bullets.length && !/^[•\-*▪◦]/.test(line)) {
        if (currentJob) { result.workExperience.push(currentJob); }
        const date = endsWithDateMatch[1];
        const title = line.replace(/((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4}))\s*$/i, '').trim().replace(/[\s|–—-]+$/, '').trim();
        currentJob = { title, date, company: '', location: '', bullets: [] };
      } else if (currentJob && !currentJob.company && (line.includes('|') || line.includes(' - ') || line.includes('–') || line.toLowerCase().includes('full time') || line.toLowerCase().includes('remote'))) {
        const parts = line.split(/[|–-]/).map(p => p.trim()).filter(Boolean);
        if (parts.length >= 2) {
          currentJob.company = parts[0];
          currentJob.location = parts.slice(1).join(' - ');
        } else {
          currentJob.location = line;
        }
      } else if (currentJob && !currentJob.location && (line.toLowerCase().includes('full time') || line.toLowerCase().includes('part time') || line.toLowerCase().includes('remote') || line.includes('India') || line.includes('USA'))) {
        currentJob.location = line;
      } else if (currentJob) {
        const txt = line.replace(/^[•\-*▪◦\s]+/, '').trim();
        if (txt) currentJob.bullets.push(txt);
      } else {
        currentJob = { title: line, date: '', company: '', location: '', bullets: [] };
      }
    } else if (currentSection === 'EDUCATION') {
      const fullRangeRx = /((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4})[-–\s\u2013\u2014]*(?:Present|\d{4}|Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)?)/i;
      const dm = line.match(fullRangeRx);

      if (currentEdu && !currentEdu.date && dm) {
        currentEdu.date = dm[1];
        const degreeText = line.replace(fullRangeRx, '').trim();
        if (degreeText) currentEdu.degree = degreeText;
      } else if (dm) {
        if (currentEdu) { result.education.push(currentEdu); }
        const date = dm[1];
        const inst = line.replace(fullRangeRx, '').trim().replace(/[\s|–—-]+$/, '').trim();
        currentEdu = { institution: inst, date, degree: '' };
      } else if (currentEdu) {
        currentEdu.degree = currentEdu.degree ? currentEdu.degree + ' ' + line : line;
      } else {
        currentEdu = { institution: line, date: '', degree: '' };
      }
    } else {
      const cust = result.customSections.find(s => s.title === currentSection);
      if (cust) cust.content.push(line);
    }
  }

  if (currentJob) result.workExperience.push(currentJob);
  if (currentEdu) result.education.push(currentEdu);

  // Filter out dangling empty skill categories
  result.skills = result.skills.filter(s => s.items && s.items.trim());

  // Infer skills if empty
  if (result.skills.length === 0) {
    result.skills = [
      { category: 'Programming Languages', items: 'Go (Golang), Python, TypeScript, SQL, Shell Scripting' },
      { category: 'Cloud, Automation & Infrastructure', items: 'AWS, Azure, GCP, Docker, Kubernetes, Helm, Terraform, CI/CD, Ollama' },
      { category: 'Data & Streaming', items: 'Redis, MongoDB, PostgreSQL, Google Pub/Sub, NATS, Kafka, gRPC, REST APIs, WebSocket' },
      { category: 'Practices & Tools', items: 'Test-Driven Development (TDD), Microservices Architecture, Event-Driven Architecture, System Design, Git, Mocha, SLB OSDU Data Platform' }
    ];
  }

  return result;
}

export function convertStructuredToText(sr: any): string {
  if (!sr) return '';
  const lines: string[] = [];

  if (sr.name) lines.push(sr.name.trim());
  if (sr.title) lines.push(sr.title.trim());
  if (sr.contact_items && sr.contact_items.length > 0) {
    lines.push(sr.contact_items.join(' | '));
  } else if (sr.contactItems && sr.contactItems.length > 0) {
    lines.push(sr.contactItems.join(' | '));
  }

  lines.push('');

  if (sr.summary) {
    lines.push('PROFESSIONAL SUMMARY');
    lines.push(sr.summary.trim());
    lines.push('');
  }

  const work = sr.work_experience || sr.workExperience || [];
  if (work.length > 0) {
    lines.push('WORK EXPERIENCE');
    work.forEach((job: any) => {
      const titleLine = `${job.title || ''} ${job.date ? job.date : ''}`.trim();
      if (titleLine) lines.push(titleLine);
      if (job.company || job.location) {
        const compLine = [job.company, job.location].filter(Boolean).join(' | ');
        lines.push(compLine);
      }
      if (job.bullets && job.bullets.length > 0) {
        job.bullets.forEach((b: string) => {
          if (b && b.trim()) lines.push(`• ${b.trim()}`);
        });
      }
      if (job.tech_stack || job.techStack) {
        lines.push(`Technologies / Skills Used : ${job.tech_stack || job.techStack}`);
      }
      lines.push('');
    });
  }

  const edu = sr.education || [];
  if (edu.length > 0) {
    lines.push('EDUCATION');
    edu.forEach((e: any) => {
      const instLine = `${e.institution || ''} ${e.date ? e.date : ''}`.trim();
      if (instLine) lines.push(instLine);
      if (e.degree) lines.push(e.degree.trim());
      lines.push('');
    });
  }

  const skills = sr.skills || [];
  if (skills.length > 0) {
    lines.push('SKILLS');
    skills.forEach((s: any) => {
      if (s.category && s.items) {
        lines.push(`${s.category.trim()} : ${s.items.trim()}`);
      }
    });
    lines.push('');
  }

  return lines.join('\n').trim();
}

File: backend/internal/parser/resume_parser.go
// Package parser provides a modular, regex-safe fallback parser that converts
// plain-text resume content into a models.StructuredResume.
//
// Design goals:
//   - All compiled regexes live as package-level variables so they are compiled
//     exactly once at program startup, not on every call.
//   - Each parsing stage (header, contact, section body) is a discrete
//     function, making each piece independently testable.
//   - No hard-coded personal data in fallback skill lists – instead an
//     empty but valid slice is returned when no skills are detected.
package parser

import (
	"regexp"
	"remotehunter/internal/models"
	"strings"
)

// ── Compiled regexes (initialized once at startup) ────────────────────────────

var (
	// reCite strips citation markers injected by AI responses.
	reCite = regexp.MustCompile(`\[cite:\s*\d+\]`)

	// reEmail matches a standard e-mail address.
	reEmail = regexp.MustCompile(`[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}`)

	// rePhone matches international and local phone number formats.
	rePhone = regexp.MustCompile(`(?:\+?\d{1,3}[\s\-]?)?\(?\d{3,5}\)?[\s\-]?\d{3,5}[\s\-]?\d{3,5}`)

	// reJobTitle matches common professional title suffixes following optional
	// prefixes such as "Senior", "Lead", "Staff", etc.
	reJobTitle = regexp.MustCompile(`(?i)^([A-Za-z\s,\(\)\/\-]+?(?:Engineer|Developer|Architect|Manager|Lead|Specialist|Consultant|Scientist|Designer|Analyst|Programmer))`)

	// reDateRange matches a full date range like "Jan 2020 – Present" or
	// "2018 - 2021".
	reDateRange = regexp.MustCompile(`(?i)((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4})[-–\s\x{2013}\x{2014}]+(?:Present|\d{4}|Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec))`)

	// reTrailingDate matches a trailing single month/year token used to stitch
	// together two-line date ranges.
	reTrailingDate = regexp.MustCompile(`(?i)((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec|\d{4}))\s*$`)

	// reSeparator matches common separator characters used inside job lines.
	reSeparator = regexp.MustCompile(`[|–—\-]+`)

	// knownSectionHeaders is the canonical set of uppercase section names
	// used as sentinels in the state machine.
	knownSectionHeaders = []string{
		"PROFESSIONAL SUMMARY", "SUMMARY",
		"WORK EXPERIENCES", "WORK EXPERIENCE", "EXPERIENCE",
		"SKILLS", "TECHNICAL SKILLS",
		"EDUCATIONS", "EDUCATION",
		"PROJECTS", "CERTIFICATIONS",
	}

	// bulletPrefixes is the set of characters used to introduce bullet points.
	bulletPrefixCutset = "•-*▪◦ "
)

// ── Public API ────────────────────────────────────────────────────────────────

// ParseText converts raw plain-text resume content into a StructuredResume.
// It is safe to call concurrently because all state is local to the function.
func ParseText(text string) *models.StructuredResume {
	// Strip citation markers left by AI systems.
	text = reCite.ReplaceAllString(text, "")

	lines := nonEmptyLines(text)
	res := &models.StructuredResume{
		ContactItems:   []string{},
		WorkExperience: []models.JobExperience{},
		Education:      []models.EducationItem{},
		Skills:         []models.SkillCategory{},
	}
	if len(lines) == 0 {
		return res
	}

	idx := parseHeader(lines, res)
	parseBody(lines[idx:], res)
	finalizeSkills(res)

	return res
}

// IsHTMLContent reports whether text already looks like an HTML document so
// the caller can skip parsing and use the string directly.
func IsHTMLContent(text string) bool {
	trimmed := strings.TrimSpace(text)
	if len(trimmed) == 0 {
		return false
	}
	// Check first character for '<' or look for the word "html" in the first
	// 50 characters (without slicing past the end of the string).
	prefix := trimmed
	if len(prefix) > 50 {
		prefix = prefix[:50]
	}
	return strings.HasPrefix(trimmed, "<") || strings.Contains(strings.ToLower(prefix), "html")
}

// ── Header parsing ────────────────────────────────────────────────────────────

// parseHeader extracts the candidate's name, title, and contact items from the
// top of the resume (lines before the first recognised section header).
// It returns the index of the first line that was NOT consumed.
func parseHeader(lines []string, res *models.StructuredResume) int {
	if len(lines) == 0 {
		return 0
	}

	// First non-empty line is always the candidate name.
	res.Name = lines[0]
	idx := 1

	for idx < len(lines) {
		line := lines[idx]
		if isSectionHeader(line) {
			break
		}

		hasEmail := reEmail.MatchString(line)
		hasPhone := rePhone.MatchString(line)

		if res.Title == "" && !hasEmail && !hasPhone && !strings.Contains(line, "|") && len(line) < 80 {
			// Short, clean line with no contact markers → professional title.
			res.Title = line
		} else {
			// Harvest contact tokens from this line.
			workLine := line

			// Try inline title extraction when title is still missing.
			if res.Title == "" {
				if m := reJobTitle.FindStringSubmatch(workLine); len(m) > 1 {
					res.Title = strings.TrimSpace(m[1])
					workLine = strings.TrimSpace(workLine[len(m[0]):])
				}
			}

			if em := reEmail.FindString(workLine); em != "" {
				appendUnique(&res.ContactItems, em)
				workLine = strings.TrimSpace(strings.Replace(workLine, em, " ", 1))
			}
			if pm := rePhone.FindString(workLine); pm != "" {
				appendUnique(&res.ContactItems, pm)
				workLine = strings.TrimSpace(strings.Replace(workLine, pm, " ", 1))
			}
			for _, part := range strings.Split(workLine, "|") {
				if p := strings.TrimSpace(part); p != "" {
					appendUnique(&res.ContactItems, p)
				}
			}
		}
		idx++
	}
	return idx
}

// ── Body parsing (section state machine) ─────────────────────────────────────

type section int

const (
	sectionUnknown section = iota
	sectionSummary
	sectionSkills
	sectionWork
	sectionEducation
)

// parseBody drives the section-level state machine over the remaining lines.
func parseBody(lines []string, res *models.StructuredResume) {
	current := sectionUnknown
	var currentJob *models.JobExperience
	var currentEdu *models.EducationItem

	flush := func() {
		if currentJob != nil {
			res.WorkExperience = append(res.WorkExperience, *currentJob)
			currentJob = nil
		}
		if currentEdu != nil {
			res.Education = append(res.Education, *currentEdu)
			currentEdu = nil
		}
	}

	for _, line := range lines {
		if isSectionHeader(line) {
			flush()
			current = classifySection(line)
			continue
		}

		switch current {
		case sectionSummary:
			parseSummaryLine(line, res)
		case sectionSkills:
			parseSkillLine(line, res)
		case sectionWork:
			parseWorkLine(line, res, &currentJob)
		case sectionEducation:
			parseEducationLine(line, res, &currentEdu)
		}
	}
	flush()
}

// ── Section classifiers ───────────────────────────────────────────────────────

func classifySection(line string) section {
	upper := strings.ToUpper(strings.TrimSpace(line))
	switch {
	case strings.Contains(upper, "SUMMARY"):
		return sectionSummary
	case strings.Contains(upper, "SKILL"):
		return sectionSkills
	case strings.Contains(upper, "WORK") || strings.Contains(upper, "EXPERIENCE"):
		return sectionWork
	case strings.Contains(upper, "EDU"):
		return sectionEducation
	default:
		return sectionUnknown
	}
}

// isSectionHeader returns true when line is a known or all-caps section header.
func isSectionHeader(line string) bool {
	upper := strings.ToUpper(strings.TrimSpace(line))
	for _, h := range knownSectionHeaders {
		if upper == h {
			return true
		}
	}
	// Heuristic: short, all-uppercase line with no separators → likely a header.
	return len(line) < 40 &&
		line == upper &&
		!strings.Contains(line, "|") &&
		!strings.Contains(line, "@") &&
		len(line) > 3
}

// ── Per-section line parsers ──────────────────────────────────────────────────

func parseSummaryLine(line string, res *models.StructuredResume) {
	if res.Summary == "" {
		res.Summary = line
	} else {
		res.Summary += " " + line
	}
}

// parseSkillLine adds a skill category parsed from line into res.
func parseSkillLine(line string, res *models.StructuredResume) {
	if colon := strings.Index(line, ":"); colon >= 0 {
		cat := strings.TrimSpace(line[:colon])
		items := strings.TrimSpace(line[colon+1:])
		res.Skills = append(res.Skills, models.SkillCategory{Category: cat, Items: items})
		return
	}
	txt := strings.TrimLeft(line, bulletPrefixCutset)
	if txt == "" {
		return
	}
	// Append to the last category when its Items field is still empty.
	if n := len(res.Skills); n > 0 && res.Skills[n-1].Items == "" {
		res.Skills[n-1].Items = txt
		return
	}
	res.Skills = append(res.Skills, models.SkillCategory{Category: "Key Skills", Items: txt})
}

func parseWorkLine(line string, res *models.StructuredResume, cur **models.JobExperience) {
	// Technology stack line.
	if strings.HasPrefix(strings.ToLower(line), "technologies") {
		if *cur != nil {
			if i := strings.Index(line, ":"); i >= 0 {
				(*cur).TechStack = strings.TrimSpace(line[i+1:])
			} else {
				(*cur).TechStack = line
			}
		}
		return
	}

	fullMatch := reDateRange.FindStringSubmatch(line)
	trailMatch := reTrailingDate.FindStringSubmatch(line)

	// Stitch together a two-line date range: first line ended with a start
	// date, this line begins with the end date.
	if *cur != nil && (*cur).Date != "" &&
		!strings.Contains((*cur).Date, "-") &&
		!strings.Contains((*cur).Date, "Present") &&
		len(trailMatch) > 1 {

		endDate := trailMatch[1]
		(*cur).Date += " - " + endDate
		comp := strings.TrimSpace(reTrailingDate.ReplaceAllString(line, ""))
		if comp != "" {
			(*cur).Company = comp
		}
		return
	}

	// New job with a full date range on the same line.
	if len(fullMatch) > 0 && len(line) < 130 && !strings.HasPrefix(line, "•") {
		if *cur != nil {
			res.WorkExperience = append(res.WorkExperience, **cur)
		}
		*cur = &models.JobExperience{
			Title:   strings.TrimSpace(reSeparator.ReplaceAllString(reDateRange.ReplaceAllString(line, ""), "")),
			Date:    fullMatch[1],
			Bullets: []string{},
		}
		return
	}

	// New job with only a trailing date token.
	if len(trailMatch) > 0 && len(line) < 100 &&
		(*cur == nil || len((*cur).Bullets) == 0) &&
		!strings.HasPrefix(line, "•") {

		if *cur != nil {
			res.WorkExperience = append(res.WorkExperience, **cur)
		}
		*cur = &models.JobExperience{
			Title:   strings.TrimSpace(reSeparator.ReplaceAllString(reTrailingDate.ReplaceAllString(line, ""), "")),
			Date:    trailMatch[1],
			Bullets: []string{},
		}
		return
	}

	// Company / location line under an open job entry.
	if *cur != nil && (*cur).Company == "" &&
		(strings.Contains(line, "|") ||
			strings.Contains(line, " - ") ||
			strings.Contains(line, "–") ||
			strings.Contains(strings.ToLower(line), "full time") ||
			strings.Contains(strings.ToLower(line), "remote")) {

		parts := strings.FieldsFunc(line, func(r rune) bool {
			return r == '|' || r == '-' || r == '–'
		})
		if len(parts) >= 2 {
			(*cur).Company = strings.TrimSpace(parts[0])
			(*cur).Location = strings.TrimSpace(parts[1])
		} else {
			(*cur).Location = line
		}
		return
	}

	// Supplementary location line.
	if *cur != nil && (*cur).Location == "" &&
		(strings.Contains(strings.ToLower(line), "full time") ||
			strings.Contains(strings.ToLower(line), "remote") ||
			strings.Contains(line, "India")) {
		(*cur).Location = line
		return
	}

	// Bullet point under open job.
	if *cur != nil {
		if txt := strings.TrimLeft(line, bulletPrefixCutset); txt != "" {
			(*cur).Bullets = append((*cur).Bullets, txt)
		}
		return
	}

	// Orphaned line with no open job — start a new job with title only.
	*cur = &models.JobExperience{Title: line, Bullets: []string{}}
}

func parseEducationLine(line string, res *models.StructuredResume, cur **models.EducationItem) {
	dm := reDateRange.FindStringSubmatch(line)

	// Attach date to an open education entry.
	if *cur != nil && (*cur).Date == "" && len(dm) > 0 {
		(*cur).Date = dm[1]
		if deg := strings.TrimSpace(reDateRange.ReplaceAllString(line, "")); deg != "" {
			(*cur).Degree = deg
		}
		return
	}

	// New education entry.
	if len(dm) > 0 {
		if *cur != nil {
			res.Education = append(res.Education, **cur)
		}
		*cur = &models.EducationItem{
			Institution: strings.TrimSpace(reSeparator.ReplaceAllString(reDateRange.ReplaceAllString(line, ""), "")),
			Date:        dm[1],
		}
		return
	}

	// Degree detail under open education entry.
	if *cur != nil {
		if (*cur).Degree == "" {
			(*cur).Degree = line
		} else {
			(*cur).Degree += " " + line
		}
		return
	}

	// Institution line with no date yet.
	*cur = &models.EducationItem{Institution: line}
}

// ── Post-processing ───────────────────────────────────────────────────────────

// finalizeSkills removes skill categories with empty Items and, when no skills
// were detected at all, leaves the slice empty so the caller can decide how to
// handle the gap (rather than injecting hard-coded personal data).
func finalizeSkills(res *models.StructuredResume) {
	var valid []models.SkillCategory
	for _, s := range res.Skills {
		if strings.TrimSpace(s.Items) != "" {
			valid = append(valid, s)
		}
	}
	if valid == nil {
		valid = []models.SkillCategory{}
	}
	res.Skills = valid
}

// ── Utilities ─────────────────────────────────────────────────────────────────

// nonEmptyLines splits text on newlines and discards blank lines.
func nonEmptyLines(text string) []string {
	raw := strings.Split(text, "\n")
	out := make([]string, 0, len(raw))
	for _, l := range raw {
		if t := strings.TrimSpace(l); t != "" {
			out = append(out, t)
		}
	}
	return out
}

// appendUnique appends s to *slice only when s is not already present.
func appendUnique(slice *[]string, s string) {
	for _, item := range *slice {
		if item == s {
			return
		}
	}
	*slice = append(*slice, s)
}
